### ISSUE TITLE
Enum nonmembers not supported

### ISSUE THREAD

**Issue by Gouvernathor** [microsoft/pyright#7595]
## Environment data

-   Pylance version: v2024.3.2
-   OS and version: Windows 10
-   Python version (& distribution if applicable, e.g. Anaconda): cpython 3.12

## Code Snippet

```python
class Relationship(enum.Flag):
    # technical
    NONE = 0
    SELF = 1

    # direct
    PARENT = enum.auto()
    CHILD = enum.auto()
    SPOUSE = enum.auto()

    # testing methods {Relationship : (FamilyMember, FamilyMember) -> bool}
    # the order matters : the first matching relation will be preferred between two persons
    # all relationships should be there, except for None
    _testers: ClassVar[dict[Relationship, Callable[[FamilyMember, FamilyMember], bool]]] = enum.nonmember({})

def _relationship_tester(relationship: Relationship, /):
    def decorator(func):
        Relationship._testers[relationship] = func
        return func
    return decorator
```

## Repro Steps

The `enum.nonmember({})` is underlined and shows the error `Expression of type "nonmember[dict[Any, Any]]" cannot be assigned to declared type "dict[Any, Any]"
"nonmember[dict[Any, Any]]" is incompatible with "dict[Any, Any]"Pylance(reportAssignmentType)`.

**Comment by erictraut** [comment]
The above code sample isn't self-contained. It refers to a symbol `FamilyMember`, which isn't defined anywhere in the sample. I presume this is a simple class?

I'm not able to repro any errors in either pylance (v2024.3.2) or pyright (1.1.356). I don't think there's a bug here. 

Code sample in [pyright playground](https://pyright-play.net/?code=MQAgDgngTglg5gCwC4C4QGcmwMZILABQhMAtmAPZRIgCmAdgK4mEBmU5JISEYMdcIUhSogAwgEMANpPEAjSTQA0YmenQA1cVEKFsq9CABi4kjEkQAsjRKyaUNADonOgnvFqQAJRoykMcnToCDBgABT0TA6GMnAAlCiEIEkgAHIA8ikAoiAAvCAADInJAMqZADKGuSAAjC7JIAAKAIKemSkAKlURJA7iDEjkobFFSaIAEgCSZQAiXYw9fQNDIyDFDWkAqqVzkYuDw0QE9Ug0mHboaKL6mlAA2iv1ACYwuLcARN6%2B-oHBYG-KEmkcgUt1uxlM5isNjsynBZks1lsUAAuspZORyJJkciVsidj06AESIi7KEAN4AXwOhEeNBYICgPnEfgCQRCAH0TmcoKFGV9Wb80J9md82WBlAB6eIrWn02nYSjMyihFgMOjYS5SGTyGiguGQklQWEmeFQpGokDozHI6VHerJYUsn4hBxck5QdC3PkigUhPF5VXqh7JRlIBhQOggQPYOohmhhiMgeWKgbaAhAA)

```python
# pyright: strict

import enum
from typing import Callable, ClassVar

class FamilyMember: ...

class Relationship(enum.Flag):
    NONE = 0
    SELF = 1

    PARENT = enum.auto()
    CHILD = enum.auto()
    SPOUSE = enum.auto()

    testers: ClassVar[
        dict["Relationship", Callable[[FamilyMember, FamilyMember], bool]]
    ] = enum.nonmember({})

def relationship_tester(relationship: Relationship, /):
    def decorator(func: Callable[[FamilyMember, FamilyMember], bool]):
        Relationship.testers[relationship] = func
        return func

    return decorator
```

I'm guessing this is a configuration issue. Please verify the version of pylance that you're using. If the problem persists, try to repro with a clean Python 3.12 environment.

**Comment by Gouvernathor** [comment]
The full code is [here](https://github.com/Gouvernathor/family/blob/main/family.py), and while the pyright playground's share link doesn't work for some reason (I assume the code is too long), and it fails to import the `util` submodule in the playground because of course it can't, the error I described is also very well present.

**Comment by erictraut** [comment]
I'm not able to reproduce any errors for the code you have linked. As I said, this is probably a configuration issue — likely something wrong with your Python environment such as a module named `enum` that is overriding the typeshed stub of the same name. Make sure you try to repro with a fresh virtual environment and default settings.

**Comment by Gouvernathor** [comment]
I tried by copy-pasting my code inside the pyright playground, so unless _it_ has a configuration as well, it's a problem with the checker.

**Comment by Gouvernathor** [comment]
I forgot : I'm also seeing it on two different VSCode install on as many computers (in addition to the pyright playground, that is). I'm having a hard time seeing how it can be a configuration issue.

**Comment by Gouvernathor** [comment]
@erictraut did you find the configuration issue which makes your version not behave like the pyright playground ?

**Comment by erictraut** [comment]
No, I'm not able to repro the problem you're seeing. The code that you posted at the top of this issue isn't self-contained and is incomplete. I posted my attempt to make the code complete, and this type checks without errors in both VS Code and the [pyright playground](https://pyright-play.net/?code=MQAgDgngTglg5gCwC4C4QGcmwMZILABQhMAtmAPZRIgCmAdgK4mEBmU5JISEYMdcIUhSogAwgEMANpPEAjSTQA0YmenQA1cVEKFsq9CABi4kjEkQAsjRKyaUNADonOgnvFqQAJRoykMcnToCDBgABT0TA6GMnAAlCiEIEkgAHIA8ikAoiAAvCAADInJAMqZADKGuSAAjC7JIAAKAIKemSkAKlURJA7iDEjkobFFSaIAEgCSZQAiXYw9fQNDIyDFDWkAqqVzkYuDw0QE9Ug0mHboaKL6mlAA2iv1ACYwuLcARN6%2B-oHBYG-KEmkcgUt1uxlM5isNjsynBZks1lsUAAuspZORyJJkciVsidj06AESIi7KEAN4AXwOhEeNBYICgPnEfgCQRCAH0TmcoKFGV9Wb80J9md82WBlAB6eIrWn02nYSjMyihFgMOjYS5SGTyGiguGQklQWEmeFQpGokDozHI6VHerJYUsn4hBxck5QdC3PkigUhPF5VXqh7JRlIBhQOggQPYOohmhhiMgeWKgbaAhAA).

I don't plan to spend any further time on this unless you're able to provide repro steps, preferably with a minimal, self-contained code sample.

**Comment by rchiodo** [comment]
I can't reproduce it either using your github sample. 

This line here, right? (There are other errors in the file):

![image](https://github.com/microsoft/pyright/assets/19672699/7726a22d-77e0-4cfd-b722-6acb7108e91c)

**Comment by Gouvernathor** [comment]
That's weird, now it doesn't work on the playground, I wisk I had taken a screenshot of it.
Surprisingly, in my VSCode, the error disappeared tonight... only to reappear sometime later, after I changed only comments at the end of the file. Maybe it will reappear on the playground later too, who knows ?
@rchiodo yes, that's the line, though I don't get the error on `func` that you're having, neither in VSCode or on the playground. The only one I have is the import of the `util` submodule when only giving playground the one file.

**Comment by Gouvernathor** [comment]
I did more tests, trying to minimize the required code. The error keeps disappearing and reappearing.
It seems it has more chances of appearing after I add new code to the file, and more chances to disappear after I remove bits of the code.
I'm not sure what to do.
